runtime/runtime/src/actions.rs

```diff
@@ -624,6 +627,193 @@ pub(crate) fn action_add_key(
     Ok(())
 }
 
+pub(crate) fn apply_delegate_action(
+    state_update: &mut TrieUpdate,
+    apply_state: &ApplyState,
+    action_receipt: &ActionReceipt,
+    sender_id: &AccountId,
+    signed_delegate_action: &SignedDelegateAction,
+    result: &mut ActionResult,
+) -> Result<(), RuntimeError> {
+    let delegate_action = &signed_delegate_action.delegate_action;
+
+    if !signed_delegate_action.verify() {
+        result.result = Err(ActionErrorKind::DelegateActionInvalidSignature.into());
+        return Ok(());
+    }
+    if apply_state.block_height > delegate_action.max_block_height {
+        result.result = Err(ActionErrorKind::DelegateActionExpired.into());
+        return Ok(());
+    }
+    if delegate_action.sender_id.as_str() != sender_id.as_str() {
+        result.result = Err(ActionErrorKind::DelegateActionSenderDoesNotMatchTxReceiver {
+            sender_id: delegate_action.sender_id.clone(),
+            receiver_id: sender_id.clone(),
+        }
+        .into());
+        return Ok(());
+    }
+
+    validate_delegate_action_key(state_update, apply_state, delegate_action, result)?;
+    if result.result.is_err() {
+        return Ok(());
+    }
+
+    let actions = delegate_action.get_actions();
+
+    let new_receipt = Receipt::new_delegate_actions(
+        &action_receipt.signer_id,
+        sender_id,
+        &delegate_action.receiver_id,
+        &actions,
+        &action_receipt.signer_public_key,
+        action_receipt.gas_price,
+    );
+
+    let prepaid_send_fees = total_prepaid_send_fees(
+        &apply_state.config.transaction_costs,
+        &action_receipt.actions,
+        apply_state.current_protocol_version,
+    )?;
+    let required_gas = receipt_required_gas(apply_state, &new_receipt)?;
+    result.gas_used += required_gas;
+    result.gas_used += prepaid_send_fees;
+    result.gas_burnt += prepaid_send_fees;
```

Comment:
nit: We usually use `safe_add_gas` to explicitly handle integer overflows. I know this is impossible to overflow with today's gas settings but for consistency and ease of mind, it would be better to use that here as well.